### PR TITLE
src/test/osd: create ceph_test_rados_io_sequence testing tool

### PR DESCRIPTION
**Pull request by JonBailey1993**
src/test/osd: Create ceph_test_rados_io_sequence tool

Application creates IO sequences using seeded random numbers. IO sequences are varied in IO offset, length, IO order and data patterns to try stress erasure coding specifically and detect defects earlier in the development cycle.

We feel the creation of this as a new tool is benificial for several reasons:
* Existing test tools either generate completely random I/O (with different offsets and lengths) or test very uniform I/O (either read/write whole object or read/writes with fixed length). This is very inefficient (relies on brute force) for testing boundary conditions. The I/O sequence test tool generates sequences of I/Os to specifically test boundary conditions
* Existing test tools only generate RADOS requests that contain a single read or write request. Recent regressions found in EC show that clients (e.g. RBD with LUKs) can generate RADOS requests that contain multiple reads or writes. This tool will sometimes generate requests with 2 or 3 I/O operations
* Quality of data patterns when writing to objects and validation of data for reading is not as great as desired in existing test tools. This means that code bugs where the wrong part of an object is read/written may not be detected. This tool creates more sophisticated data patterns and keeps a model of the object to track the expected contents of the object.
* For erasure coding there are many choices of erasure code profile, this tool can create pools and exercise multiple different profiles in parallel to speed up testing.

The test tool is located in src/test/osd, as the new tool is used explicitly for testing the erasure coding implementation of OSDs. A lot of the logic for generating IO Sequences has been stored in a new location, io_exerciser, located in src/common. The thinking behind storing the code here is we feel it is useful to share between other applications. It is abstracted to a level it shouldn't be necissairy to have to be only used with only the new ceph_test_rados_io_sequence application and we would like to be able to integrate with other tools such as ceph_test_rados or possibly rados bench as well in future PRs to enhance testing done with those tools.

Signed-off-by: Jon Bailey <jonathan.bailey1@ibm.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/common/io_exerciser/Model.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace ceph {
+  namespace io_exerciser {
+
+    class Model
+    {
+    protected:
+      int num_io;
+      const std::string oid;
```
Comment (ronen-fr) [inline src/common/io_exerciser/Model.h:30]: consider not using 'const' members, as this prevents moving.

File: src/common/io_exerciser/Model.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      uint64_t block_size;
+
+    public:
+      Model(const std::string oid, uint64_t block_size);
```
Comment (ronen-fr) [inline src/common/io_exerciser/Model.h:34]: 'const string' as a parameter spec does not make good sense.

Comment (ronen-fr) [reply src/common/io_exerciser/Model.h:34]: Basically - for all but string, accept as const T&.
If expecting a string - and planning to keep a copy - accept a string by value.
(and - if the environment is right - string_view)

File: src/common/io_exerciser/Model.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ */
+
+namespace ceph {
+  namespace io_exerciser {
+
+    class Model
+    {
+    protected:
+      int num_io;
```
Comment (ronen-fr) [inline src/common/io_exerciser/Model.h:29]: int num_io{0}, as you only have one ctor, and that's what it sets

File: src/common/io_exerciser/ObjectModel.cc [hunk 4]
```diff
@@ -0,0 +1,162 @@
+#include "ObjectModel.h"
+
+ceph::io_exerciser::ObjectModel::ObjectModel(const std::string oid, uint64_t block_size, int seed) :
```
Comment (ronen-fr) [inline src/common/io_exerciser/ObjectModel.cc:3]: This is a cpp file, so it makes sense to shorten the 'ceph::io_ex...'.
E.g.
using ObjectModel = ceph.....::objectmodel

File: src/common/io_exerciser/Model.h (hunk 1 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/Model.h:30]: Was not aware of this nuance. Will update.

File: src/common/io_exerciser/Model.h (hunk 2 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/Model.h:34]: Was intending to put const T& but missed the &, got spot.

It sounds like by what you are saying however is that I should pass the string by value still, just without the const, is this correct? (I presume then move into the member variable instead of copy if this is what you are suggesting).

File: src/common/io_exerciser/Model.h (hunk 3 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/Model.h:29]: Will update. Is this just a readability improvement? I presume the compiled code that comes from this will still be the same?

File: src/common/io_exerciser/ObjectModel.cc (hunk 4 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/ObjectModel.cc:3]: Sure, I can 'using' the namespaces in the cpp files in a few places throughout the pr if this is the preferred method, as they are fairly long in places.

File: src/common/io_exerciser/ObjectModel.cc [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    break;
+
+  case CREATE:
+    ceph_assert(!created);
+    ceph_assert(reads.empty());
+    ceph_assert(writes.empty());
+    created = true;
+    contents.resize(op.length1);
+    for (uint64_t i = 0; i < contents.size(); i++) {
+      contents[i] = rng();
```
[the remark this answers is not in the document]
Comment (JonBailey1993) [reply src/common/io_exerciser/ObjectModel.cc:65]: Am not aware of the std::generate function. Will investigate and update.

File: src/common/io_exerciser/Model.h (hunk 2 above)
Comment (ronen-fr) [reply src/common/io_exerciser/Model.h:34]: Basically - we can usually count on the passed string to exist in the caller context until the called function returns. If that function does not stash a ref to the passed string somewhere (which will require it to outlive the caller),
than the best way to 'receive' it is as a string_view if possible (4 years ago - it would have been 'const string&').
If  it does require the string to outlive the caller: we'll need a copy (or move the string in, if the caller does not need it anymore). If the latter - that would usually require passing std::move(s).

File: src/common/io_exerciser/Model.h (hunk 3 above)
Comment (ronen-fr) [reply src/common/io_exerciser/Model.h:29]: Not just readability. It also guarantees that if anyone ever adds a 2'nd ctor - that initialization would not be forgotten.
And yes - the resulting code is the same.

File: src/common/io_exerciser/Model.h (hunk 2 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/Model.h:34]: Ok, it sounds like because this is a constructor and it's stashing the oid in this case then I'm requiring the string to outlive the caller. I also want the oid to exist in the caller context after I construct this object potentially, so I do not want to std::move it out of the caller so it can still be used there - which leaves me with the option of passing it by copy.

In this case I would do this as passing it as "const std::string**&** oid", and then using the copy constructor of std::string in the initialiser list to copy this referenced string into the Model object so that the copy stored in Model can outlive the original string if necessary, using ", oid(oid)" in the initialiser list.

Have I understood this correctly?

Apologies for the questions, I'm just trying to nail down my understanding of the theory here for my knowledge.

Comment (ronen-fr) [reply src/common/io_exerciser/Model.h:34]: Yes. Sorry for nagging about that issue.

File: src/common/io_exerciser/IoOp.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * class IoOp
+ *   Stores details for an I/O operation. Generated by IoSequences
+ *   and applied by IoExerciser's
+ */
+
+namespace ceph {
+  namespace io_exerciser {
+
+    enum OpType {
```
Comment (ronen-fr) [inline src/common/io_exerciser/IoOp.h:20]: If possible - pls use class enums for all new code.

File: src/common/io_exerciser/IoOp.cc [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  offset1(offset1), length1(length1),
+  offset2(offset2), length2(length2),
+  offset3(offset3), length3(length3)
+{
+  
+}
+
+std::string IoOp::value_to_string(uint64_t v) const
+{
+  if (v < 1024 || (v % 1024) != 0) {
```
Comment (ronen-fr) [inline src/common/io_exerciser/IoOp.cc:19]: Note that if performance was an issue (which I expect it is not here): to_string() is usually much slower than the alternatives (e.g. to_chars).

File: src/common/io_exerciser/IoOp.cc [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+IoOp::IoOp( OpType op,
+            uint64_t offset1, uint64_t length1,
+            uint64_t offset2, uint64_t length2,
+            uint64_t offset3, uint64_t length3) :
+  op(op),
+  offset1(offset1), length1(length1),
+  offset2(offset2), length2(length2),
+  offset3(offset3), length3(length3)
+{
+
```
Comment (ronen-fr) [inline src/common/io_exerciser/IoOp.cc:14]: 2 extra \b's

File: src/common/io_exerciser/DataGenerator.cc [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+    std::memcpy(buffer + block_size - remainingBytes, &rand1, remainingBytes);
+  }
+
+  return ceph::bufferptr(buffer, block_size);
+}
+
+void SeededRandomGenerator::generate_data(uint64_t offset, uint64_t length,
+                                          bufferlist& retlist)
+{
+  ceph_assert(retlist.length() == 0);
```
Comment (ronen-fr) [inline src/common/io_exerciser/DataGenerator.cc:146]: so - if you are sure that 'relist' is out-only - it would probably be better
to change the signature to:

bufferlist generate...(uint64_t offset, uint64_t length)

File: src/common/io_exerciser/DataGenerator.h [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+          uint64_t m_uniqueId = 0;
+
+          static DataGenerationSingleton m_singletonInstance;
+          static std::mutex m_mutex;
+
+        public:
+          static const DataGenerationSingleton&
+            createSpecificInstance(uint64_t unique_id);
+          static const DataGenerationSingleton& getInstance();
+          static const uint64_t getUniqueId();
```
Comment (ronen-fr) [inline src/common/io_exerciser/DataGenerator.h:84]: no point in the 'const' on a returned uint. 
More important: note that the decision to use a singleton for the DataGenerator
object means, for example, that you have at least two more static methods.
Can the DG instance be injected to its users, instead of using a singleton?

File: src/common/io_exerciser/DataGenerator.cc [hunk 11]
```diff
@@ -0,0 +1,713 @@
...
+    }
+    break;
+    
+    case ErrorType::SEED_MISMATCH:
+    {
+      SeedBytes read_seed = readSeed((block_offset - read_offset), bufferlist);
+
+      if (m_model.get_seed_offsets(read_seed).size() == 0)
+      {
+        ss << "Data (Seed) mismatch detected at block " << block_offset
```
Comment (ronen-fr) [inline src/common/io_exerciser/DataGenerator.cc:347]: I'd recommend using std::format() for all new code, instead of chains of streams

**Comment by ronen-fr** [comment]
nice work!

**Comment by ronen-fr** [comment]
LGTM apart from some style comments.
Marking as 'approved', but would appreciate you checking the comments.

**Review by ronen-fr — APPROVED** [review]
see comments

**Comment by JonBailey1993** [comment]
> LGTM apart from some style comments. Marking as 'approved', but would appreciate you checking the comments.

Will check through comments this morning and update code in places. Thanks for taking the time to review this.

File: src/common/io_exerciser/IoOp.h (hunk 6 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/IoOp.h]: Done

File: src/common/io_exerciser/IoOp.cc (hunk 8 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/IoOp.cc]: Removed the whitespace here

File: src/common/io_exerciser/DataGenerator.cc (hunk 9 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/DataGenerator.cc]: Not sure why I did it like this originally - have updated.

File: src/common/io_exerciser/DataGenerator.h (hunk 10 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/DataGenerator.h]: Have refactored out the use of the singleton and instead generate the unique ID optionally with the HeaderedSeededRandomGenerator, as this was the only class that I intended to make use of this

File: src/common/io_exerciser/DataGenerator.cc (hunk 11 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/DataGenerator.cc]: Have updated to make use of the fmt::format library you mentioned later on rather than stringstreams.

File: src/common/io_exerciser/RadosIo.cc [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+  while (outstanding_io > count) {
+    cond.wait(l);
+  }
+}
+
+void RadosIo::allow_ec_overwrites(bool allow)
+{
+  int rc;
+  bufferlist inbl, outbl;
+  std::string cmdstr =
```
[the remark this answers is not in the document]
Comment (JonBailey1993) [reply src/common/io_exerciser/RadosIo.cc:64]: I have not updated this with fmt::format because I am planning on adding more json commands in another update separate to this merge request and would prefer to deal with this section in that request. I do agree with you this could be easier to read and I would like to spend the time when I create that PR to update the json creation and make it common code between the commands rather than how it is created here, along with investigating creating them through json libraries included in the ceph source already.

File: src/common/io_exerciser/DataGenerator.cc [hunk 13]
```diff
@@ -0,0 +1,713 @@
...
+#include <iostream>
+#include <stdexcept>
+
+#define dout_subsys ceph_subsys_rados
+#define dout_context g_ceph_context
+
+using DataGenerator = ceph::io_exerciser::data_generation::DataGenerator;
+using SeededRandomGenerator = ceph::io_exerciser::data_generation
+                                ::SeededRandomGenerator;
+using HeaderedSeededRandomGenerator = ceph::io_exerciser::data_generation
```
[the remark this answers is not in the document]
Comment (JonBailey1993) [reply src/common/io_exerciser/DataGenerator.cc:23]: That sounds smart and would increase the readability and I may still look at that in the future, but for now I'd like to leave this as is. It sounds like a neat way of doing things, but I would like to make that a separate delivery if I do it and just to try get this PR merged as is for now if your happy with this being my call.

**Comment by JonBailey1993** [comment]
> see comments

Have addressed all comments now, either by resolving in code or with a reply to the comment

**Comment by JonBailey1993** [comment]
**ceph API tests** check failed.

Failed during build step installing dashboard nodeenv:
`[1772/1776] dashboard nodeenv is being installed
FAILED: src/pybind/mgr/dashboard/frontend/node-env/bin/npm /home/jenkins-build/build/workspace/ceph-api/build/src/pybind/mgr/dashboard/frontend/node-env/bin/npm`

Looking further down the logs, looks like connection was interrupted during download of node
` * Install prebuilt node (20.13.1) .Incomplete read while reading from https://nodejs.org/download/release/v20.13.1/node-v20.13.1-linux-x64.tar.gz - IncompleteRead(3186688 bytes read, 43534552 more expected)
Incomplete read while reading from https://nodejs.org/download/release/v20.13.1/node-v20.13.1-linux-x64.tar.gz - IncompleteRead(2531328 bytes read, 44189912 more expected)
Incomplete read while reading from https://nodejs.org/download/release/v20.13.1/node-v20.13.1-linux-x64.tar.gz - IncompleteRead(14393344 bytes read, 32327896 more expected)`

Will retrigger **ceph API tests** as looks to have been a network glitch during building from my inspection here.

**Comment by JonBailey1993** [comment]
jenkins test api

**Comment by JonBailey1993** [comment]
**make check (arm64)** check failed.

Failure reason looks similar to the last test
`FAILED: src/pybind/mgr/dashboard/frontend/CMakeFiles/mgr-dashboard-nodeenv /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/CMakeFiles/mgr-dashboard-nodeenv`

Same step was above and in progress:
`[2306/2698] dashboard nodeenv is being installed`

Failure reason looks to be the same, with extra text in the middle due to presumably another thread starting a different build step while the previous one was working towards the failure
` * Install prebuilt node (20.13.1) .Incomplete read while reading from https://nodejs.org/download/release/v20.13.1/node-v20.13.1-linux-arm64.tar.gz - IncompleteRead(1908736 bytes read, 44521098 more expected)
..... done.
 * Appending data to /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/node-env/bin/activate
 * Appending data to /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/node-env/bin/activate.fish
[2307/2698] cd /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend && . /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/node-env/bin/activate && npm config set cache /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/node-env/.npm --userconfig /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/node-env/.npmrc && deactivate
FAILED: src/pybind/mgr/dashboard/frontend/CMakeFiles/mgr-dashboard-nodeenv /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/build/src/pybind/mgr/dashboard/frontend/CMakeFiles/mgr-dashboard-nodeenv `

I believe this is the same network glitch on the build system and will retrigger the **make check (arm64)** test

**Comment by JonBailey1993** [comment]
jenkins test make check arm64

**Comment by JonBailey1993** [comment]
jenkins test api

**Comment by JonBailey1993** [comment]
jenkins test make check arm64

**Comment by JonBailey1993** [comment]
`301/301 Test #262: unittest-transaction-manager ..............***Timeout 7201.31 sec`
One unit test for arm64 hit timeout and was aborted.

**Comment by JonBailey1993** [comment]
jenkins test make check arm64

File: src/common/io_exerciser/IoOp.cc (hunk 7 above)
Comment (JonBailey1993) [reply src/common/io_exerciser/IoOp.cc:19]: I have not updated this and will take the performance hit for now. That is useful to know for future reference though when I am looking at code with higher performance requirements though, thank you.

File: src/common/io_exerciser/RadosIo.cc [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+  offset1(offset1), length1(length1),
+  offset2(offset2), length2(length2),
+  offset3(offset3), length3(length3)
+{
+
+}
+
+bool RadosIo::readyForIoOp(IoOp &op)
+{
+  ceph_assert(lock.is_locked_by_me()); //Must be called with lock held
```
Comment (benhanokh) [inline src/common/io_exerciser/RadosIo.cc:84]: Line fails build complaining about missing function is_locked_by_me().
Build fails when doing a DEBUG build

Comment (JonBailey1993) [reply src/common/io_exerciser/RadosIo.cc:84]: I have a PR open to fix this here: https://github.com/ceph/ceph/pull/60211, as this code has now unfortunately merged with this line in.

However I noticed the opposite when opening the PR - that the build fails when doing a RELEASE/RelWithDebInfo build and succeeds when doing a DEBUG build. Can I confirm the build type you are doing please? If it is debug, are you able to provide any more information?

Comment (benhanokh) [reply src/common/io_exerciser/RadosIo.cc:84]: I was doing DEBUG build (which activates the assert code) and failed with a complain about  non-existing member function is_locked_by_me() for BOOST::lock.

I comment out the assert line in my code and now it builds fine
